### PR TITLE
tech detection: Parallelize loading of tech files

### PR DESCRIPTION
**Pull request by kingthorin**
## Overview
Parallelize loading of tech files for performance reasons.

## Related Issues
N/A

## Checklist
- [NA] Update help
- [X] Update changelog
- [X] Run `./gradlew spotlessApply` for code formatting
- [NA] Write tests
- [X] Check code coverage
- [X] Sign-off commits
- [X] Squash commits
- [X] Use a descriptive title

### CODE REVIEW

**Comment by thc202** [comment]
There's a typo in Parallelize both in PR title and commit message.

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java [hunk 1]
```diff
@@ -66,12 +72,33 @@ public WappalyzerJsonParser() {
...
     WappalyzerData parse(String categories, List<String> technologies, boolean createIcons) {
         LOGGER.info("Starting to parse Tech Detecction technologies.");
+        Instant start = Instant.now();
         WappalyzerData wappalyzerData = new WappalyzerData();
         parseCategories(wappalyzerData, getStringResource(categories));
-        technologies.forEach(
-                path -> parseJson(wappalyzerData, getStringResource(path), createIcons));
+        // Process the files/paths in parallel
+        ExecutorService executor =
+                Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() - 1);
```
Comment (thc202) [inline addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:80]: - Worth checking with `commonlib`'s `Constants`.
 - Better shutdown the executor once no longer needed.
 - Name the threads, if there's a problem it's easier to identify the what.

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java [hunk 2]
```diff
@@ -66,12 +72,33 @@ public WappalyzerJsonParser() {
...
+                                                        parseJson(
+                                                                wappalyzerData,
+                                                                getStringResource(path),
+                                                                createIcons),
+                                                executor))
+                        .collect(Collectors.toList());
+        // Get all the finished tasks
+        @SuppressWarnings("unused")
+        List<?> results =
+                futures.stream().map(CompletableFuture::join).collect(Collectors.toList());
```
Comment (thc202) [inline addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:96]: Why collect twice?

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java [hunk 3]
```diff
@@ -109,7 +136,7 @@ private void parseCategories(WappalyzerData wappalyzerData, String jsonStr) {
     }
 
     @SuppressWarnings("unchecked")
-    private void parseJson(WappalyzerData wappalyzerData, String jsonStr, boolean createIcons) {
+    private boolean parseJson(WappalyzerData wappalyzerData, String jsonStr, boolean createIcons) {
```
Comment (thc202) [inline addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:139]: I'd suggest keep void and just change the task to return null.

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java (hunk 2 above)
Comment (kingthorin) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:96]: That was based on examples I was looking at all of which had a start and then collect finished. But maybe it isn't needed. Will test further.

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java (hunk 3 above)
Comment (kingthorin) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:139]: `List<CompletableFuture<null>>` isn't accepted.

Comment (kingthorin) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:139]: May have got it.

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java (hunk 2 above)
Comment (kingthorin) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:101]: Leaving "futures" as a stream is actually markedly slower for some reason.

Comment (kingthorin) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:101]: Now just one collect.

Comment (thc202) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java]: Maybe it's the use of Constants, which uses more threads.

Comment (thc202) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java]: It should not need to collect, just do the forEach directly.

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java [hunk 4]
```diff
@@ -66,12 +75,35 @@ public WappalyzerJsonParser() {
...
+                                        CompletableFuture.runAsync(
+                                                () ->
+                                                        parseJson(
+                                                                wappalyzerData,
+                                                                getStringResource(path),
+                                                                createIcons),
+                                                executor))
+                        .collect(Collectors.toList());
+        // Finished the tasks
+        futures.forEach(x -> x.join());
```
Comment (thc202) [inline addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java:100]: Use the method reference as before.

**Comment by kingthorin** [comment]
FYI I did try breaking a pattern and make sure that still fails the test. Shall I add a file that purposefully contains a broken entry and assert that the error list after parsing isn't empty?

**Comment by thc202** [comment]
Always good to cover all paths (on purpose).

It would indeed be more correct (though less likely to cause any problems there since there are not that many issues at a single time).

File: addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java (hunk 2 above)
Comment (thc202) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java]: Bump.

Comment (thc202) [reply addOns/wappalyzer/src/main/java/org/zaproxy/zap/extension/wappalyzer/WappalyzerJsonParser.java]: For the record, having them split performs better.

**Comment by thc202** [comment]
Thank you!

**Review by ricekot — APPROVED** [review]
This is awesome, thank you!

**Pull request merged**